### DISCUSSION TITLE
Optimization: constant and dead logic cleaning while keeping hierarchy

### DISCUSSION THREAD

**Discussion by xtofalex** [YosysHQ/yosys#3835] [Q&A] [answered]
Hi everyone,
I would like to know if there is, in Yosys a combination of commands and switches that allows constant and dead logic optimizations, while keeping hierarchy and module interfaces.
Thanks a lot !
Christophe

**Comment by nakengelhardt** [comment, marked as answer]
`opt` does both of these optimizations, but it is run separately on each module. So as long as you keep the hierarchy, it is not possible to do optimization across module boundaries.

**Reply by xtofalex** [reply]
Thanks a lot for your answer.
Can you please confirm that applying uniquification does not make a difference ? This is what I'm experiencing now, but just want to make sure I'm not missing something.

**Reply by nakengelhardt** [reply]
Indeed, `uniquify` will make a separate copy of the module for each instance, but they will still have the same module boundaries as the original module, so optimization is only done on the module contents - it doesn't know anything about the instantiation context.

**Reply by xtofalex** [reply]
Thanks again. This clarifies everything.
